[net] Make cs_inventory nonrecursive

jnewbery opened the pull request:
- Remove PushBlockInventory() and PushBlockHash(). These are one-line functions that can easy be inlined into the calling code. Doing so also allows us to eliminate the one place that cs_inventory is recursively locked.
- Make cs_inventory a nonrecursive mutex
- Remove a redundant TRY_LOCK of cs_inventory when deleting CNode.

jnewbery committed:
[net processing] Remove PushBlockInventory and PushBlockHash

PushBlockInventory() and PushBlockHash() are functions that can
be replaced with single-line statements. This also eliminates
the single place that cs_inventory is taken recursively.

jnewbery committed:
[net] Make cs_inventory a non-recursive mutex

cs_inventory is never taken recursively. Make it a non-recursive mutex.

jnewbery committed:
[net] Don't try to take cs_inventory before deleting CNode

The TRY_LOCK(cs_inventory) in DisconnectNodes() is taken after the CNode
object has been removed from vNodes and when the CNode's nRefCount is
zero.

The only other places that cs_inventory can be taken are:

- In ProcessMessages() or SendMessages(), when the CNode's nRefCount
must be >0 (see ThreadMessageHandler(), where the refcount is
incremented before calling ProcessMessages() and SendMessages()).
- In a ForEachNode() lambda in PeerLogicValidation::UpdatedBlockTip().
ForEachNode() locks cs_vNodes and calls the function on the CNode
objects in vNodes.

Therefore, cs_inventory is never locked by another thread when the
TRY_LOCK(cs_inventory) is reached in DisconnectNodes(). Since the
only purpose of this TRY_LOCK is to ensure that the lock is not
taken by another thread, this always succeeds. Remove the check.

jnewbery commented:
@hebasto : you seem to be on a mutex-cleaning spree. Perhaps you'd be interested in this?

hebasto commented:
Concept ACK. Made a link here from the tracking issue #19303.

hebasto commented on src/net.cpp:
```diff
@@ -1103,12 +1103,9 @@ void CConnman::DisconnectNodes()
-                        if (lockSend) {
-                            fDelete = true;
-                        }
+                    TRY_LOCK(pnode->cs_vSend, lockSend);
+                    if (lockSend) {
+                        fDelete = true;
```
From commit message
> Therefore, cs_inventory can never be locked when the TRY_LOCK(cs_inventory) is reached in DisconnectNodes(). Remove the TRY_LOCK.

I understand that `if (lockInv)` is never true, and the code block `TRY_LOCK(pnode->cs_vSend, lockSend); ...` is dead in master. This commit makes this code always reachable, right?

MarcoFalke replied:
I think the comment says "Therefore, cs_inventory can never be locked *by another thread* ...", so it is always true (success).

jnewbery replied:
Sorry, you're right that this commit message is unclear. Marco is correct that I meant "cs_inventory can never be locked by another thread". I'll update the commit message.

jnewbery replied:
I've made the commit log clearer.

hebasto reviewed (commented):
ACK eaced81501f035193523d1abde4569dfc73ae327, I've verified that mutex does not get locked recursively. Also tested on Linux Mint 20 (x86_64) with the `double_lock_detected()` function from #19337.

For safety in the future I suggest to merge this PR on top of the #19337 which preserves UB in case of any recursive locking.

hebasto commented:
nit: s/`cs_inventory`/`m_inventory_mutex`/

jnewbery commented:
> nit: s/cs_inventory/m_inventory_mutex

I agree! I'm planning to move this mutex into net processing where it belongs. I'll do the rename at the same time. See https://github.com/jnewbery/bitcoin/commits/2020-06-cs-main-split

laanwj commented on src/net_processing.cpp:
```diff
@@ -1330,7 +1330,7 @@ void PeerLogicValidation::UpdatedBlockTip(const CBlockIndex *pindexNew, const CB
             if (nNewHeight > (pnode->nStartingHeight != -1 ? pnode->nStartingHeight - 2000 : 0)) {
                 for (const uint256& hash : reverse_iterate(vHashes)) {
-                    pnode->PushBlockHash(hash);
+                    WITH_LOCK(pnode->cs_inventory, pnode->vBlockHashesToAnnounce.push_back(hash));
```
When doing this anyway, woulid it make sense to move this lock outside the loop? Re-taking it for every `node × every block` seems overkill.

jnewbery replied:
Good idea. Done.

promag replied:
e335e4fcc881260906fb561f23a1b5017ed04d91

The suggestion was done in 6db15a8b79597a0f311490acd89ee50db298c6e8, but IMO should be in e335e4fcc881260906fb561f23a1b5017ed04d91.

jnewbery replied:
Done.

hebasto commented:
Does the recent [change](https://github.com/bitcoin/bitcoin/compare/eaced81501f035193523d1abde4569dfc73ae327..6db15a8b79597a0f311490acd89ee50db298c6e8) decrease concurrentness of the code? It seems like a premature optimization a bit, no?

jnewbery commented:
> Does the recent change decrease concurrentness of the code?

I think it's fine to expand the lock scope:

- it's _extremely_ unlikely that this'll increase contention for this lock. `UpdatedBlockTip()` is only called once every ten minutes on average, it's very unlikely that it'll push two block hashes onto the `vBlockHashesToAnnounce` vector (only if there's a re-org), and the time it takes to push an extra hash onto the vector is tiny.
- in the very unlikely case we do block the other thread, it's probably an improvement. Rather than sending one hash in a headers message to the peer, and then another hash in a second headers message, we'll send both in the same message.

jnewbery commented:
I've moved the lock scope change to the correct commit, and also moved it one line up, outside the surrounding if block. It doesn't make any difference here, but makes a commit in a future branch tidier.

sipa commented:
utACK e8a2822119233ade0de84f791a9e92918a3d6896

hebasto reviewed (approved):
re-ACK e8a2822119233ade0de84f791a9e92918a3d6896

MarcoFalke commented:
ACK e8a2822119233ade0de84f791a9e92918a3d6896 🍬

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK e8a2822119233ade0de84f791a9e92918a3d6896 🍬
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgLcQv/SQmH5S8+LYeoatDRRklzvqBa7vDSX+1JFkyauaibxrGa2j0CT5pmbrQi
VBQSMZNrfC3F7g0hydzbplkkDzTciuj0rJGK6ScTmPLJA1X1ytCCoOjncCB+hz5O
L250NVZCJRsbESdtIzqfmJZcl/ANYbxeM9ZF+sgVFvG5tq+OMmHy1zWN6Yejxnxh
vryK7Ke16Mny2QLmrgohTQTyYicHzX0wrKDV2v7J1DQL6TaN4nVAsmPV7eKbIAI6
wTPl8msWKfknfrGU2SEurFEdfKN0IapV8KxM+trbjKpakIGcF3sFaP60cHzf1haG
HjfckGlwx9hiYWs6d6RwM9mGI+t8k4UkdPbxrAwjW/OnK0jAw6Zp/NOJK7VX19Em
DcoKZXMHmrdNvIV5sxxXkGJtEQJsknuyIyjmbOB9HezJG+obnSNGuMvdqF7WLV20
UQpV+WhV07LlYz4akXZ/SGsDvRJkFzIskaIc4u2tKVVHM/HIHjsrB7Vy2DTd8A4J
XLhWoDVu
=9k86
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `19ea9d208bf92e8a00c6a74b291941159f29e27a9ad985a83fc50fb9e7eeb14a  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010819ea9d208bf92e8a00c6a74b291941159f29e27a9ad985a83fc50fb9e7eeb14af0108df50901dbda9abb6da7ebacc3340c3308fff0102fd60bd06ce468d455f1b5a4105cf06108f1045f0624f4f0080e222e4787965af90083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0106dcf23eaf62794540a9b5c3f34cee70e08f1045f0624f4f0082527e637c980e8150083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010b561711d9277ebacb3f41eba5075c00808f1045f0624f5f008b398429856ae78fb0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010e760ef17026c64cb1aa6dd772c2d0ded08f1045f0624f4f008ac0ec51240f25e420083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>
